### ISSUE TITLE
certificate is not yet valid

### ISSUE THREAD

**Issue by pai-da-xin** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#915]
Hi,Dear experts,
* An error occurred while I was running the master program, namely "certificate is not yet valid". The printed log is as follows:
> [KVS Master] Using trickleICE by default
[KVS Master] Created signaling channel ScaryTestChannel
[KVS Master] Finished setting audio and video handlers
[KVS Master] KVS WebRTC initialization completed successfully
1970-01-01 02:06:24 INFO    createSignalingClientSync(): Creating Signaling Client Sync
1970-01-01 02:06:24 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000001, Next state: 0x0000000000000002
1970-01-01 02:06:24 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000002, Next state: 0x0000000000000004
[1970/01/01 02:06:28:1157] E: SSL error: certificate is not yet valid (preverify_ok=0;err=9;depth=3)
1970-01-01 02:06:28 WARN    lwsHttpCallbackRoutine(): Client connection failed. Connection error string: client connect failed
1970-01-01 02:06:28 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000004, Next state: 0x0000000000000010
[1970/01/01 02:06:30:0529] E: SSL error: certificate is not yet valid (preverify_ok=0;err=9;depth=3)
1970-01-01 02:06:30 WARN    lwsHttpCallbackRoutine(): Client connection failed. Connection error string: client connect failed
1970-01-01 02:06:30 ERROR   signalingCacheSaveToFile(): operation returned status code: 0x00000002
1970-01-01 02:06:30 WARN    getChannelEndpoint(): Failed to save signaling cache to file
1970-01-01 02:06:30 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000010, Next state: 0x0000000000000020
1970-01-01 02:06:30 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000020, Next state: 0x0000000000000040
[KVS Master] Signaling client created successfully
1970-01-01 02:06:30 INFO    signalingClientConnectSync(): Signaling Client Connect Sync
1970-01-01 02:06:30 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000040, Next state: 0x0000000000000080
1970-01-01 02:06:30 ERROR   connectSignalingChannelLws(): operation returned status code: 0x0000000c
1970-01-01 02:06:30 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000080, Next state: 0x0000000000000002
1970-01-01 02:06:30 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000002, Next state: 0x0000000000000004
1970-01-01 02:06:30 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000004, Next state: 0x0000000000000010
1970-01-01 02:06:30 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000010, Next state: 0x0000000000000020
1970-01-01 02:06:30 DEBUG   refreshIceConfigurationCallback(): Refreshing the ICE Server Configuration
1970-01-01 02:06:55 ERROR   signalingConnectSync(): operation returned status code: 0x0000000f
[KVS Master] signalingClientConnectSync(): operation returned status code: 0x0000000f 
[KVS Master] Terminated with status code 0x0000000f 
[KVS Master] Cleaning up....
1970-01-01 02:06:55 DEBUG   logSignalingClientStats(): Signaling client connection duration: 0 ms
1970-01-01 02:06:55 DEBUG   logSignalingClientStats(): Number of signaling client API errors: 1
1970-01-01 02:06:55 DEBUG   logSignalingClientStats(): Number of runtime errors in the session: 0
1970-01-01 02:06:55 DEBUG   logSignalingClientStats(): Signaling client uptime: 0 ms
1970-01-01 02:06:55 DEBUG   logSignalingClientStats(): Control Plane API call latency: 265 ms
1970-01-01 02:06:55 DEBUG   logSignalingClientStats(): Data Plane API call latency: 0 ms
1970-01-01 02:06:55 INFO    freeSignalingClient(): Freeing Signaling Client
[KVS Master] Cleanup done

Is this a network problem, a cert.pem certificate problem or something?

**Comment by MushMal** [comment]
Check the system clock and adjust it

**Comment by pai-da-xin** [comment]
How to check the system clock and adjust it. Is it to adjust the computer clock?

**Comment by MushMal** [comment]
That would depend on your platform. Some platforms use NTP to sync the clock. This is beyond the SDK and its dependent on your platform and the architecture.

Try to debug your platform to understand why your clock is drifted back.

Please resolve this issue if you don’t have any further questions related to KVS itself

**Comment by pai-da-xin** [comment]
@MushMal 
* When I run the master, the printing time is the network time. Why is it still not working? The function connectSignalingChannelLws returns status code 0x0000000c. Is this still a clock problem? The printed log is as follows:
> [KVS Master] Signaling client created successfully
2020-11-03 07:35:35 INFO    signalingClientConnectSync(): Signaling Client Connect Sync
2020-11-03 07:35:35 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000040, Next state: 0x0000000000000080
2020-11-03 07:35:35 ERROR   connectSignalingChannelLws(): operation returned status code: 0x0000000c
2020-11-03 07:35:35 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000080, Next state: 0x0000000000000002
2020-11-03 07:35:35 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000002, Next state: 0x0000000000000004
2020-11-03 07:35:35 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000004, Next state: 0x0000000000000010
2020-11-03 07:35:35 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000010, Next state: 0x0000000000000020
2020-11-03 07:35:35 DEBUG   refreshIceConfigurationCallback(): Refreshing the ICE Server Configuration
2020-11-03 07:36:00 ERROR   signalingConnectSync(): operation returned status code: 0x0000000f
[KVS Master] signalingClientConnectSync(): operation returned status code: 0x0000000f 
[KVS Master] Terminated with status code 0x0000000f 
[KVS Master] Cleaning up....

**Comment by MushMal** [comment]
This is an entirely different issue. In this case you are hitting an internal error defined here
https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/common/include/com/amazonaws/kinesis/video/common/CommonDefs.h#L454


I have no idea what could have caused it but here are some recommendations 

Cut a separate issue targeting that particular error case
Do some of your own debugging - the samples need two line change to use IoT credentials provider instead of the static ones. Make no other modifications. 
Run in verbose level instead of debug log level 
Try to get as much info as possible which would help you root cause the issue
Run the stock sample on a standard platform (Linux Mac Windows) with static credentials provider. 
Run the same with IoT

Only after that cut an issue if you still have problems and include the detailed information with full log to avoid multiple round trips with absolutely no results.

Please close this issue

**Comment by pai-da-xin** [comment]
@MushMal 
* When I run master, there was a problem with the ice server configuration, What causes this? 

The printed log is as follows:
> [KVS Master] Signaling client created successfully
2020-11-05 03:28:56 INFO    signalingClientConnectSync(): Signaling Client Connect Sync
2020-11-05 03:28:56 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000040, Next state: 0x0000000000000080
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 8 - 'Connecting'
2020-11-05 03:28:56 ERROR   connectSignalingChannelLws(): operation returned status code: 0x0000000c
2020-11-05 03:28:56 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000080, Next state: 0x0000000000000002
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2020-11-05 03:28:56 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000002, Next state: 0x0000000000000004
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 3 - 'Describe Channel'
2020-11-05 03:28:56 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000004, Next state: 0x0000000000000010
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 5 - 'Get Channel Endpoint'
2020-11-05 03:28:56 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000010, Next state: 0x0000000000000020
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:28:56 DEBUG   refreshIceConfigurationCallback(): Refreshing the ICE Server Configuration
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:28:56 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:28:57 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:28:59 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:29:02 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:29:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2020-11-05 03:29:21 ERROR   signalingConnectSync(): operation returned status code: 0x0000000f
[KVS Master] signalingClientConnectSync(): operation returned status code: 0x0000000f

**Comment by MushMal** [comment]
As I mentioned earlier, you are hitting this line

2020-11-05 03:28:56 ERROR connectSignalingChannelLws(): operation returned status code: 0x0000000c

0x0000000c is STATUS_INTERNAL_ERROR which in this case is returned in 

https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Signaling/LwsApiCalls.c#L1233

This is just a sanity check validating that the static structure has been properly initialized.

Under normal execution, you should never hit this. If you hit this then it's likely due to an earlier memory corruption or you've modified the code.

I don't see how this can be related to the network or the clock itself. 

My recommendation stands.

* Make sure you use stock latest commit from master branch
* Run the sample without any modification
* Run the sample on a well-known platform that we have tested on - for example Linux on RPi or Mac OS
* Make your modifications on the standard platform
* Run the stock application with absolutely no modifications on your platform


We won't be able to help you this way unless you do some debugging on your own as we can't reproduce this. 

The original issue was cut and root caused to skewed platform clock (which was fairly obvious from the error message) but we are iterating over and over with issues which are unrelated to the original root cause. Moreover, we can't seem to get more details to help you drive this forward. Please follow the steps I highlighted above. If you still have any issues unrelated to certificate is not yet valid then please cut a separate issue and include detailed description of your scenario, the expectation, the observed behavior, the changes you might have made, your platform info and verbose logs to get some help.

I am resolving this issue as it's been root caused.
